### DISCUSSION TITLE
Connection pool with using PostgresqlConnection. (ClassCastException io.r2dbc.pool.PooledConnection )

### DISCUSSION THREAD

**Discussion by szbenceg** [pgjdbc/r2dbc-postgresql#592] [Q&A] [answered]
How can I get a connection from the pool and use it as a PostgresqlConnection?

I would like to call a prepared statement like this: 
In this case if I understand correctly, `databaseClient.getConnectionFactory().create()` takes one connection from the pool. (I have also tested it with this statement `SELECT * FROM pg_stat_activity WHERE datname = 'chat';`)
```
  return Mono.from(databaseClient.getConnectionFactory().create())
                .flatMapMany(c -> {
                    PostgresqlConnection connection = (PostgresqlConnection) c;
                    var statement = connection
                            .createStatement("call select_conversations($1, null)")
                            .bind("$1", Parameters.in(PostgresqlObjectId.UUID, UUID.fromString(userId)))
                            .execute()
                            .flatMap(result -> {
```

But in this case I get this exception: 
`java.lang.ClassCastException: class io.r2dbc.pool.PooledConnection cannot be cast to class io.r2dbc.postgresql.api.PostgresqlConnection (io.r2dbc.pool.PooledConnection and io.r2dbc.postgresql.api.PostgresqlConnection are in unnamed module of loader 'app')` 

Is it possible to create a `PostgresqlConnection` from a `io.r2dbc.pool.PooledConnection` somehow?

The only workaround I found was to create a new connection on the fly like this: 
```
public static ConnectionFactory fromUrl() {
    return ConnectionFactories.get(dbConnectionUrl);
}

return Mono.from(DatabaseConfiguration.fromUrl().create())
                .flatMapMany(c -> {
                    PostgresqlConnection connection = (PostgresqlConnection) c;
                    var statement = connection
                            .createStatement("call select_conversations($1, null)")
                            
```

But in this case what is the point of having a connection pool if I cannot use it?

Probably I misunderstand something, and it is just some misconfiguration. 

Thank you for your help in advance.

**Comment by mp911de** [comment, marked as answer]
You can unwrap the native connection via `Wrapped.unwrap(…)` method. The pooled connection object implements [`io.r2dbc.spi.Wrapped`](https://r2dbc.io/spec/1.0.0.RELEASE/spec/html/#wrapped).

**Reply by szbenceg** [reply]
Thank you. 🙂
